cli/types/index.d.ts

```diff
@@ -174,757 +182,739 @@ declare namespace Cypress {
+    contains<K extends keyof HTMLElementTagNameMap>(selector: K, text: string | number | RegExp, options?: Partial<Loggable & Timeoutable>): Chainable<JQuery<HTMLElementTagNameMap[K]>>    
+    contains<E extends Node = HTMLElement>(selector: string, text: string | number | RegExp, options?: Partial<Loggable & Timeoutable>): Chainable<JQuery<E>>
 
     /**
+     * Double-click a DOM element.
+     *
      * @see https://on.cypress.io/dblclick
      */
-    dblclick(options?: Loggable): Chainable
+    dblclick(options?: Partial<Loggable>): Chainable
 
     /**
+     * Set a debugger and log what the previous command yields.
+     *
      * @see https://on.cypress.io/debug
      */
-    debug(options?: Loggable): Chainable
+    debug(options?: Partial<Loggable>): Chainable<Subject>
 
     /**
      * Get the window.document of the page that is currently active.
      *
-     * @param {Loggable} [options]
-     * @returns {Chainable}
-     * @memberof Chainable
      * @see https://on.cypress.io/document
-     * @example
-     *    cy.document()
-     *      .its('contentType')
-     *      .should('eq', 'text/html')
+     * @example document
+     * ```ts
+     * cy.document()
+     *   .its('contentType')
+     *   .should('eq', 'text/html')
+     * ```
      */
-    document(options?: Loggable): Chainable
+    document(options?: Partial<Loggable>): Chainable<Document>
 
     /**
+     * Iterate through an array like structure (arrays or objects with a length property).
+     *
      * @see https://on.cypress.io/each
      */
-    each(fn: (element?: any, index?: number, $list?: any) => void): Chainable
+    each<E extends Node = HTMLElement>(fn: (element: JQuery<E>, index: number, $list: E[]) => void): Chainable<JQuery<E>> // Can't properly infer type without breaking down Chainable
 
     /**
+     * End a chain of commands
+     *
      * @see https://on.cypress.io/end
      */
-    end(): Chainable
+    end(): Chainable<null>
 
     /**
+     * Get A DOM element at a specific index in an array of elements.
+     *
      * @see https://on.cypress.io/eq
      */
-    eq(index: number, options?: LoggableTimeoutable): Chainable
+    eq<E extends Node = HTMLElement>(index: number, options?: Partial<Loggable & Timeoutable>): Chainable<JQuery<E>>
 
     /**
+     * Execute a system command.
      * @see https://on.cypress.io/exec
      */
-    exec(command: string, options?: ExecOptions): Chainable
+    exec(command: string, options?: Partial<ExecOptions>): Chainable<Exec>
 
     /**
+     * Get the DOM elements that match a specific selector. Opposite of `.not()`
+     *
      * @see https://on.cypress.io/filter
      */
-    filter(selector: string, options?: LoggableTimeoutable): Chainable
+    filter<K extends keyof HTMLElementTagNameMap>(selector: K, options?: Partial<Loggable & Timeoutable>): Chainable<JQuery<HTMLElementTagNameMap[K]>> // automatically returns the correct HTMLElement type

```

Comment:
A way of getting the right element without using a generic would be to use the `.filter` method:

```ts
cy.get('.input') // HTMLElement since we can't infer type
  .filter('input') // HTMLInputElement since 'input' is a keyof match
  .then($input => {
    $el // $ExpectType JQuery<HTMLInputElement>
  })
```

That would also always make sure you have an `<input>` or else the test will fail before it gets there. I prefer this over generic overrides since overriding generics is basically overriding the type system (possible runtime error in that case)

Reply:
I wonder if, given that info, that should be the way to get a specific element type? I don't like seeing generics in non-library code - it is just too easy to misuse.

Generics would still be required in areas we can't infer based on input (like fixture contents or results from a `.as`).

Reply:
yeah, I bet sometimes `as <...>` would still be needed, unfortunately (or fortunately if you think of the flexibility)

Reply:
I think generic overrides are necessary sometimes. I think it is a good practice to avoid the requirement. I've seen developers introduce runtime errors by incorrectly overriding them. It can be tricky because type errors can produce very difficult "type traces".